build your first basic workflow experimentation

jsundai opened the pull request:
## What does this PR do?
https://temporal-documentation-git-tutorial-layout-money-transfe-0fc0c4.preview.thundergun.io/build-your-first-basic-workflow/build-your-first-workflow 

## Notes to reviewers
files changed include new components for interaction and sdk boxed logos 
aiming for: concise + engaging, dev-to-dev tone, consistent interactions, progressive disclosure.
can include sdk box logos in the sdk tabs section in the future  
<!-- delete if n/a -->

jsundai committed:
starting place

jsundai committed:
first basic workflow python

jsundai committed:
adding components and sdk box logos

jsundai committed:
language tabs

jsundai committed:
rename for multilanguage

jsundai committed:
text edits

jsundai committed:
component clean up

jsundai committed:
adding custom language order for language tabs

jsundai committed:
title and description change

jsundai committed:
getting started section

jsundai committed:
adding in workflow and activity definition base

jsundai committed:
refining retry policy section and adding each sdk tab

jsundai committed:
add comments to retry policy code

jsundai committed:
add base of worker step

jsundai committed:
task queue base by language

jsundai committed:
add execute the workflow base

jsundai committed:
adding langauge tabs for crash test

jsundai committed:
different layout for crash test and friendlier language

jsundai committed:
experiment 2 base by sdk

jsundai committed:
restructured live bug fix section

jsundai committed:
react component for terminal animation

jsundai committed:
updating intro draft for failure simulation

jsundai committed:
update index with cards

jsundai committed:
remove redundant components, edit intro

jsundai committed:
add set up steps for project install in each language

jsundai committed:
failure simulation content editS

jsundai committed:
adding php snippets for money transfer

jsundai committed:
adding more php

jsundai committed:
add language tabs on commands

lennessyy commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,2101 @@
+
+Before starting this tutorial:
+
+- **Set up a local development environment** for developing Temporal applications
```
I feel like the prerequisites are a little unclear to those who are new to our site. I know that if users go through the quick starts if they are set, but if I'm a new user, it's not clear if I am supposed to do the setup AND the quickstarts. 

If we don't want to create a separate environment setup page, then we should just remove the two bullets and just tell them to do the quickstarts as a prerequisite.

lennessyy reviewed (commented):
This is gorgeous! Great work. My comments are aimed at user-friendliness but may make it less pretty

brianmacdonald-temporal reviewed (commented):
I feel like we could use some text explanation for each of the code blocks. Maybe Point out specifically what the workflow is calling, and how that relates to  the activities, and point out the Retry Policy details.

jsundai committed:
Update docs/build-your-first-basic-workflow/build-your-first-workflow.mdx

Co-authored-by: Milecia McG <47196133+flippedcoder@users.noreply.github.com>

jsundai committed:
Update docs/build-your-first-basic-workflow/build-your-first-workflow.mdx

Co-authored-by: Milecia McG <47196133+flippedcoder@users.noreply.github.com>

jsundai committed:
capitalization, uncollapsed, removing sdk box logos

jsundai committed:
replace static reimburse img with built in themedimage component, created and added light and dark mode reimburse diagram, max width 60

jsundai committed:
changing progress

jsundai committed:
Merge branch 'main' into tutorial-layout-money-transfer-experimentation

jsundai committed:
fix page jumping

jsundai committed:
edits

jsundai committed:
adding code highlighting and making comments visible

jsundai committed:
feedback

jsundai commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,2101 @@
+
+<RetryPolicyComparison />
+
+:::important This is a Simplified Example
```
I think the warning might be too harsh on the page

flippedcoder commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,2286 @@
+import type { PaymentDetails } from './shared';
+
+// highlight-next-line
+export async function moneyTransfer(details: PaymentDetails): Promise<string> {  // Workflow Definition: an exported async function
```
Nit: could the `// Workflow Definition: an exported async function` go above `// highlight-next-line` to keep it consistent with all the other comments?

jsundai replied:
I just removed the highlighted since I'm using snipsync now. I think code highlighting by line could be helpful though for general use around the docs.

flippedcoder commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,2286 @@
+  let depositResult: string;
+  try {
+    // highlight-next-line
+    depositResult = await deposit(details);  // Executes Activity via proxy and waits for result
```
Nit: could `// Executes Activity via proxy and waits for result` go above `// highlight-next-line` to keep it consistent with all the other comments?

flippedcoder commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,2286 @@
+import { BankingService } from './banking-client';
+
+// highlight-next-line
+export async function withdraw(details: PaymentDetails): Promise<string> {  // Activity Definition: an exported async function
```
Nit: could `// Activity Definition: an exported async function` go above `// highlight-next-line` to keep it consistent with all the other comments?

flippedcoder commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
````diff
@@ -0,0 +1,2286 @@
+
+```typescript
+// Task Queue name - used by both Worker and Workflow starter
+export const taskQueueName = 'MONEY_TRANSFER_TASK_QUEUE';
````
This is different from the code in the cloned repo.

<img width="480" height="76" alt="Screenshot 2026-01-21 at 15 10 36" src="https://github.com/user-attachments/assets/04d0de08-3f2e-4449-a6ae-78da8bb72698" />

flippedcoder commented on docs/build-your-first-basic-workflow/failure-simulation.mdx:
```diff
@@ -0,0 +1,1161 @@
+  const bank = new BankingService('bank2.example.com');
+  
+  // Comment out this working line:
+  // return await bank.deposit(details.targetAccount, details.amount, details.referenceId);
```
It might just be a formatting thing, but this looks different than [the repo](https://github.com/temporalio/money-transfer-project-template-ts/blob/main/src/activities.ts#L30) and the comment/uncomment snippets are in a different order (the good code is on top of the bug code).

flippedcoder commented on docs/build-your-first-basic-workflow/failure-simulation.mdx:
```diff
@@ -0,0 +1,1161 @@
+Try this advanced scenario of compensating transactions.
+
+
+1. **Modify the retry policy** in `workflows.py` to only retry 1 time
```
Do we only want to highlight Python here or all the languages?

jsundai committed:
Update docs/build-your-first-basic-workflow/build-your-first-workflow.mdx

Co-authored-by: Milecia McG <47196133+flippedcoder@users.noreply.github.com>

jsundai committed:
Update docs/build-your-first-basic-workflow/build-your-first-workflow.mdx

Co-authored-by: Milecia McG <47196133+flippedcoder@users.noreply.github.com>

flippedcoder commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,2286 @@
+                workflow.logger.info(
+                    f"Refund successful. Confirmation ID: {refund_output}"
+                )
+                raise deposit_err
```
This code doesn't match what's in the repo: https://github.com/temporalio/money-transfer-project-template-python/blob/main/workflows.py#L53-L61

jsundai committed:
dotnet feedback

jsundai commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
````diff
@@ -0,0 +1,2286 @@
+**shared.php** or in your Worker.php
+
+```php
+<?php
````
This will have to be added in the php code repo and same with .net and maybe the others. I'll just have to remove it for the doc for now until it's added to the code.

jsundai replied:
Removed PHP!

jsundai committed:
add full file path names and remove constant files

jsundai committed:
adding snipsync repos

jsundai committed:
snipsync 1st part for all sdks

jsundai committed:
remove php references and remove unused tutoriallangorder and icon imports and add hideunsupportedlanguage

jsundai committed:
renaming files to build your first temporal app and simulate failure

jsundai committed:
add nested tabs for activitydefinitions, implementations + interface, add snipsync where needed, merge task queue explanation into step 3 with constants tip, renumber steps

jsundai committed:
move tip, update diagram with light + dark mode, add theme aware background for diagram, simplify worker intro text, java note spacing

jsundai committed:
animated terminal cursor line fix and adding java snipsync execute workflow

jsundai committed:
remove unused img and unbold

jsundai commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,2286 @@
+      </SdkTabs>
+    </div>
+    <div style={{marginTop: '2rem', padding: '1rem', background: 'rgba(16, 185, 129, 0.1)', borderRadius: '0.5rem'}}>
+      <strong>Expected Success Output:</strong>
```
I'm not sure if I want to put too much emphasis on the multiple terminals 😅

jsundai committed:
feedback task queue note and open in new tab for cta

tlalfano commented:
I'd love to see the name of the page and in the left hand nav be something more like: 
Build a Financial Transaction Application, or Build a Financial Transaction Workflow 

I'm mainly concerned with this appearing to be 'build your first net new workflow' which it really isn't - and that's something we do want to solve. 

@jsundai

jsundai committed:
first part of tom's feedback

jsundai committed:
Restructure money transfer tutorial: shorter, cleaner, 4-section layout

- Cut from 2,248 to 1,771 lines (1,133 lines removed)
- Replaced SetupSteps side-panel pattern with inline code blocks
- Merged Steps 3+4+"Run" into a single "Run the application" section
- Removed <details open>, Introduction heading, and "Let's Recap" sub-step
- Condensed intro prose to three sentences; moved prereqs to a short callout
- Worker code now co-located with its run command per SDK tab
- Removed unused imports (SetupSteps, CodeSnippet)

jsundai committed:
refactoring tutorial

jsundai committed:
rename tutorial title

jsundai committed:
intro copy tweak and add missing tutorial components

tlalfano commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,1996 @@
+
+### Activities
+
+Activities do the real work. Each one calls an external service. If an Activity fails, Temporal retries it. The three Activities in this app: Withdraw, Deposit, and Refund.
```
This phrasing sounds a little odd, maybe 'There are three Activities' ?

tlalfano commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,1996 @@
+Start the local Temporal server. This runs the scheduler, state store, and Web UI at [localhost:8233](http://localhost:8233). The `--db-filename` flag persists Workflow state to disk, which is required for the crash recovery demo in Part 2.
+
+:::note Port conflict
+If you get an `address already in use` error, a previous process may still hold port 7233. Run `lsof -ti:7233 | xargs kill` on macOS/Linux to clear it.
```
In the sentence above, we say the UI is running on 8233, but in this sentence we talk about port conflicts with 7233, where the temporal service is running, but we have never intro'ed that info even though we are referencing 7233 throughout the sample application code.

tlalfano commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,1996 @@
+
+Start the Worker. It connects to Temporal, polls the Task Queue, and executes your Workflow and Activity code. Leave it running.
+
+:::note Java: first-run compile time
```
Not a deal breaker, but is there a way to only display this when Java is active? if its a massive lift its not worth spending a ton of time on imo

tlalfano commented on static/img/moneytransfer/webuisample.png:
This will likely need updating soon, UI colors and treatments have changed a bit.

tlalfano commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,1996 @@
+  <img src="/img/moneytransfer/webuisample.png" alt="Temporal Web UI showing a completed money transfer Workflow" style={{maxWidth: '100%', height: 'auto'}} />
+</div>
+
+Click on the Workflow in the list to see the full event history. In Part 2, you'll use that history to debug a live failure.
```
I get to the bottom of part 1, and I don't have an easy way to get to Part 2. I have to go all the way back to the top to click part 2. The footer of the page also suggests that 'Evaluate' is the next page that I should go to.

tlalfano commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,1996 @@
+
+## Scenario A: Crash recovery
+
+The goal is to show that Temporal stores Workflow state durably. State lives on the server, not in the Worker process, and survives server restarts.
```
Phrasing feels internal, perhaps 'this scenario will show that Temporal....'

tlalfano commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,1996 @@
+  { id: 'workflow', label: 'Workflow Status', value: 'RUNNING', status: 'running' }
+]} />
+
+:::note Python and .NET: move quickly
```
Is there an important reason for this? Feels like extra work for the end user while learning

tlalfano commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,1996 @@
+
+### Step 3: Fix the bug and redeploy
+
+Stop the Worker (`Ctrl+C`) and revert your change. Put the working deposit line back and remove (or re-wrap) the failing version. Then restart the Worker in Terminal 2:
```
In step 2 we have a code snippet that shows exactly what to comment/uncomment, should we do the same here? put the line back or re-wrap failing version sounds like a different instruction than step 2

tlalfano commented on docs/build-your-first-basic-workflow/build-your-first-workflow.mdx:
```diff
@@ -0,0 +1,1996 @@
+- **.NET**: Terminal 3 is still running and prints the transfer result when the Workflow completes.
+
+:::tip
+You just fixed a bug in a running application without losing state or restarting the transaction. Temporal stored every completed step, so Withdraw was not re-run. Execution resumed from the first failing Activity and completed from there.
```
I'd love to see this as an explicit step to review instead of a tip. I'm imagining an image of event history, or even timeline view where it shows very clearly that Withdraw was successful (green), but the deposit step continually retried (red).

tlalfano reviewed (commented):
Added my notes! Overall looking so good, loving this option of adding more tailored quick starts for users and can't wait to add more.

jsundai commented:
Closing this PR. Recently had a meeting on Aligning the Getting Started Paths and this can be used as a template for that future work.
